presto-hive/src/main/java/com/facebook/presto/hive/HiveStagingFileCommitter.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.hive.HdfsEnvironment.HdfsContext;
+import com.facebook.presto.hive.PartitionUpdate.StagingFileInfo;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.SchemaTableName;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_FILESYSTEM_ERROR;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HiveStagingFileCommitter
+        implements StagingFileCommitter
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final Executor renameExecutor;
+
+    public HiveStagingFileCommitter(
+            HdfsEnvironment hdfsEnvironment,
+            Executor renameExecutor)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.renameExecutor = requireNonNull(renameExecutor, "renameExecutor is null");
+    }
+
+    @Override
+    public void commitFiles(ConnectorSession session, SchemaTableName tableName, List<PartitionUpdate> fragments)
+    {
+        List<Future<?>> fileRenameFutures = new ArrayList<>();
+        for (PartitionUpdate partitionUpdate : fragments) {
+            Path path = partitionUpdate.getWritePath();
+            HdfsContext context = new HdfsContext(session, tableName.getSchemaName(), tableName.getTableName());
+            FileSystem fileSystem;
+            try {
+                fileSystem = hdfsEnvironment.getFileSystem(context, path);
+            }
+            catch (IOException e) {
+                throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error moving data files to final location. Error listing directory %s", path), e);
```

Comment:
We are not listing dirrectories yet. I would put here something more neutral, e.g.: https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java#L519